Spring 5 Kaalid & Liban

brittLiban opened the pull request:
***Link - http://146.190.131.66/***

-------------------------
- Did automated deployment and continuous integration (CI) using GitHub Actions
- Fix / refines setup scripts for improved Docker compatibility, and updated docs to reflect these enhancements. 

**Deployment & CI Automation**

* Added a GitHub Actions workflow (`.github/workflows/deploy.yml`) for automated deployment to the VM on every push to the `main` branch
* 
* Updated `README.md` to document the new automated deployment process, required repository secrets.
* Also added comprehensive documentation for the CI workflow, detailing test coverage and manual run instructions.

**Setup & Compatibility Improvements**

* Refactored all setup and update scripts (`setup.sh`, `setup-local.sh`, `setup-local-testing.sh`, `update.sh`) to use the modern `docker compose` plugin instead of the deprecated `docker-compose` 

* Improved Docker installation logic in `prepare-vm.sh` by removing conflicting packages and installing Docker components only if missing, ensuring a clean and reliable environment.

**Script Robustness & Cross-platform Support**

* Enhanced script checks in all setup scripts to separately verify the presence of `docker` and the `docker compose` plugin, providing clearer error messages and guidance for missing dependencies.

* Fixed a bug in `setup.sh` where the `SED` command assignment for Linux and macOS was incorrect, ensuring proper cross-platform compatibility

brittLiban committed:
Fix syntax error in OS type check in setup script

brittLiban committed:
The setup scripts assume a vm is going to be completely clean. Trying to make it robust and check if the system already has some dependencies and then work with whwats given to prevent uneccesary overwrites and other errors... phew

brittLiban committed:
fix setup script error handling and dependency checks

brittLiban committed:
scripts now are going to use 'docker compose' command and improve error messages for Docker requirements

brittLiban committed:
Add GitHub Actions workflow for VM deployment

brittLiban committed:
Merge pull request #1 from brittLiban/feature/autoDeploy

Feature/auto deploy

brittLiban committed:
Add workflow_dispatch trigger to deploy workflow

brittLiban committed:
Fix docker-compose command in deploy workflow

KaalidArare committed:
Added a test.yml file

KaalidArare committed:
Added backend unit testing to CI workflow

KaalidArare committed:
Added backend unit/integration testings and also frontend deployment

KaalidArare committed:
Added front/end-to-end testing to the CI workflow

KaalidArare committed:
Trigger CI rerun

KaalidArare committed:
Added a workflow_dispatch for workflow rerun

KaalidArare committed:
Retrigger CI

KaalidArare committed:
Retrigger CI

KaalidArare committed:
Run CI on feature/ci-merging

KaalidArare committed:
Updated README to include what the continuous integration runs and how to manually run it

brittLiban committed:
add initial GitHub Actions workflow for deployment

brittLiban committed:
Update README to include automated deployment instructions using GitHub Actions

brittLiban committed:
Merge pull request #2 from brittLiban/feature/ci-merging

Feature/ci merging

dannymccarragher commented on .github/workflows/test.yml:
Mixture of yml syntax. There is branches: -, and pull_request: [].

alstondsouza1 commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+
+        steps:
+        - name: Checkout code
+          uses: actions/checkout@v2 #checks the lattest version of the code, v2 for the latest maintianed v of this cmd from github
```
suggestions updating to actions/checkout@v4 for security and performance

dannymccarragher commented on .github/workflows/deploy.yml:
Good usage of Github Secrets!

alstondsouza1 commented on .github/workflows/deploy.yml:
consider adding a pre-check to verify all required secrets exist before SSH setup. This will make workflow failure messages more clear

alstondsouza1 commented on .github/workflows/test.yml:
Great fallback logic for missing test scripts. Very clean approach!

dannymccarragher commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,33 @@
+        - name: Checkout code
+          uses: actions/checkout@v2 #checks the lattest version of the code, v2 for the latest maintianed v of this cmd from github
+        
+        - name: Set up SSH
```
Naming of each step is done well. Simple but explains what is being accomplished.

alstondsouza1 commented on setup.sh:
The OSTYPE logic has duplicated “else” blocks consider simplifying this section for clarity

alstondsouza1 commented on prepare-vm.sh:
good improvement adding logic to remove conflicting Docker packages. suggestion add logging for each step to help debugging on fresh VMs

brittLiban committed:
Add RCA for Docker installation error on VM

Documented the RCA for Docker installation issues on VM, including troubleshooting steps and corrective actions taken to improve the setup script.
